File: research/aysenur/liveness-p3-frr-reduction/app/infrastructure/ml/liveness/face_usability_gate.py
"""Pre-liveness face usability gate for critical occlusion handling."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

import numpy as np

from app.domain.entities.face_landmarks import LandmarkResult
from app.infrastructure.ml.liveness.critical_region_visibility_gate import (
    CriticalRegionVisibilityGate,
)
from app.infrastructure.ml.liveness.face_quality_illumination_gate import (
    FaceQualityIlluminationGate,
)

LOW_QUALITY_CONFIRM_FRAMES = 2
OCCLUSION_CONFIRM_FRAMES = 2
NO_FACE_CONFIRM_FRAMES = 6
TEMP_CLEAR_CONFIRM_FRAMES = 2
CLEAR_CONFIRM_FRAMES = 5

_EYE_STRICT_UNRELIABLE_THRESHOLD = 0.60


@dataclass(frozen=True)
class FaceUsabilityResult:
    usable: bool
    no_face: bool
    quality_ok: bool
    occluded: bool
    occlusion_score: float
    occluded_regions: tuple[str, ...]
    visibility_scores: dict[str, float]
    region_reasons: dict[str, str]
    blocking_regions: tuple[str, ...]
    suspicious_regions: tuple[str, ...]
    quality_status: str
    quality_reason: str
    per_region_brightness: dict[str, float]
    brightness_uniformity: float
    illumination_score: float
    global_face_brightness: float
    shadow_asymmetry: float
    underexposed_regions: tuple[str, ...]
    overexposed_regions: tuple[str, ...]
    physical_occlusion_score: float
    physical_occlusion_regions: tuple[str, ...]
    physical_occlusion_reason: str
    liveness_skipped_reason: str
    reason: str
    state: str
    blocked: bool
    status_override: Optional[str]
    bbox_detected: bool
    occlusion_streak: int = 0
    quality_streak: int = 0
    clear_streak: int = 0


class FaceUsabilityGate:
    """Block liveness scoring when no usable full face is visible."""

    def __init__(
        self,
        *,
        low_quality_confirm_frames: int = LOW_QUALITY_CONFIRM_FRAMES,
        occlusion_confirm_frames: int = OCCLUSION_CONFIRM_FRAMES,
        no_face_confirm_frames: int = NO_FACE_CONFIRM_FRAMES,
        temp_clear_confirm_frames: int = TEMP_CLEAR_CONFIRM_FRAMES,
        clear_confirm_frames: int = CLEAR_CONFIRM_FRAMES,
    ) -> None:
        self._quality_gate = FaceQualityIlluminationGate()
        self._visibility_gate = CriticalRegionVisibilityGate()
        self._low_quality_confirm_frames = max(1, int(low_quality_confirm_frames))
        self._occlusion_confirm_frames = max(1, int(occlusion_confirm_frames))
        self._no_face_confirm_frames = max(self._occlusion_confirm_frames, int(no_face_confirm_frames))
        self._temp_clear_confirm_frames = max(1, int(temp_clear_confirm_frames))
        self._clear_confirm_frames = max(1, int(clear_confirm_frames))
        self.reset()

    def evaluate(
        self,
        *,
        frame: np.ndarray,
        face_bbox: Optional[tuple[int, int, int, int]],
        landmarks: Optional[LandmarkResult] = None,
        preview_details: Optional[dict[str, object]] = None,
        blur_score: Optional[float] = None,
    ) -> FaceUsabilityResult:
        if frame is None or frame.size == 0 or face_bbox is None:
            self._quality_streak = 0
            self._occlusion_streak = 0
            self._clear_streak = 0
            self._blocked = False
            self._state = "NO_FACE"
            return FaceUsabilityResult(
                usable=False,
                no_face=True,
                quality_ok=False,
                occluded=False,
                occlusion_score=0.0,
                occluded_regions=(),
                visibility_scores={},
                region_reasons={},
                blocking_regions=(),
                suspicious_regions=(),
                quality_status="-",
                quality_reason="-",
                per_region_brightness={},
                brightness_uniformity=0.0,
                illumination_score=0.0,
                global_face_brightness=0.0,
                shadow_asymmetry=0.0,
                underexposed_regions=(),
                overexposed_regions=(),
                physical_occlusion_score=0.0,
                physical_occlusion_regions=(),
                physical_occlusion_reason="-",
                liveness_skipped_reason="no_face_detected",
                reason="no_face_detected",
                state=self._state,
                blocked=True,
                status_override="NO_FACE",
                bbox_detected=False,
                occlusion_streak=0,
                quality_streak=0,
                clear_streak=0,
            )

        quality = self._quality_gate.evaluate(
            frame_bgr=frame,
            face_bounding_box=face_bbox,
            landmarks=landmarks,
        )
        visibility = self._visibility_gate.evaluate(
            frame_bgr=frame,
            face_bounding_box=face_bbox,
            landmarks=landmarks,
            preview_details=preview_details,
            blur_score=blur_score,
        )
        left_eye_score = visibility.visibility_scores.get("left_eye", 1.0)
        right_eye_score = visibility.visibility_scores.get("right_eye", 1.0)
        left_eye_visible = left_eye_score >= 0.60
        right_eye_visible = right_eye_score >= 0.60
        nose_visible = visibility.visibility_scores.get("nose", 1.0) >= 0.65
        mouth_visible = visibility.visibility_scores.get("mouth", 1.0) >= 0.65
        lower_face_visible = visibility.visibility_scores.get("lower_face", 1.0) >= 0.60
        both_eyes_unreliable = bool(
            left_eye_score < _EYE_STRICT_UNRELIABLE_THRESHOLD
            and right_eye_score < _EYE_STRICT_UNRELIABLE_THRESHOLD
        )
        derived_occluded_regions: list[str] = list(visibility.occluded_regions)
        if both_eyes_unreliable or (not left_eye_visible and not right_eye_visible):
            derived_occluded_regions.extend(["left_eye", "right_eye"])
        if not nose_visible:
            derived_occluded_regions.append("nose")
        if not mouth_visible:
            derived_occluded_regions.append("mouth")
        if not lower_face_visible:
            derived_occluded_regions.append("lower_face")
        derived_occluded_regions = list(dict.fromkeys(derived_occluded_regions))

        structural_occlusion_now = bool(
            both_eyes_unreliable
            or (not left_eye_visible and not right_eye_visible)
            or (not nose_visible)
            or (not mouth_visible and not lower_face_visible)
            or (
                visibility.occlusion_score >= 0.58
                and ((not mouth_visible) or (not nose_visible) or (not lower_face_visible))
            )
        )
        occluded_now = bool(visibility.is_critical_occluded or structural_occlusion_now)
        if occluded_now:
            self._quality_streak = 0
            self._occlusion_streak += 1
            self._clear_streak = 0
            self._blocked = True
            self._blocked_mode = "OCCLUSION"
            if self._occlusion_streak >= self._no_face_confirm_frames:
                self._entered_no_face = True
                self._state = "OCCLUDED_NO_FACE"
                status_override = "NO_FACE"
            elif self._occlusion_streak >= self._occlusion_confirm_frames:
                self._state = "OCCLUDED_CONFIRMED"
                status_override = "INSUFFICIENT_EVIDENCE"
            else:
                self._state = "OCCLUDED_PENDING"
                status_override = "INSUFFICIENT_EVIDENCE"
            return FaceUsabilityResult(
                usable=False,
                no_face=True,
                quality_ok=True,
                occluded=True,
                occlusion_score=visibility.occlusion_score,
                occluded_regions=tuple(derived_occluded_regions),
                visibility_scores=dict(visibility.visibility_scores),
                region_reasons=dict(visibility.region_reasons),
                blocking_regions=tuple(derived_occluded_regions),
                suspicious_regions=tuple(
                    region
                    for region in dict.fromkeys([*visibility.suspicious_regions, *derived_occluded_regions])
                    if region not in derived_occluded_regions
                ),
                quality_status=quality.quality_status,
                quality_reason=quality.quality_reason,
                per_region_brightness=dict(quality.per_region_brightness),
                brightness_uniformity=quality.brightness_uniformity,
                illumination_score=quality.illumination_score,
                global_face_brightness=quality.global_face_brightness,
                shadow_asymmetry=quality.shadow_asymmetry,
                underexposed_regions=quality.underexposed_regions,
                overexposed_regions=quality.overexposed_regions,
                physical_occlusion_score=visibility.occlusion_score,
                physical_occlusion_regions=tuple(derived_occluded_regions),
                physical_occlusion_reason=(
                    visibility.reason
                    if visibility.is_critical_occluded
                    else "structural_face_region_occluded"
                ),
                liveness_skipped_reason="critical_face_region_occluded",
                reason="critical_face_region_occluded",
                state=self._state,
                blocked=True,
                status_override=status_override,
                bbox_detected=True,
                occlusion_streak=self._occlusion_streak,
                quality_streak=0,
                clear_streak=0,
            )

        self._occlusion_streak = 0
        self._quality_streak = 0
        if self._blocked:
            self._clear_streak += 1
            required_clear_frames = (
                self._clear_confirm_frames
                if self._entered_no_face or self._blocked_mode == "LOW_QUALITY"
                else self._temp_clear_confirm_frames
            )
            if self._clear_streak < required_clear_frames:
                self._state = "RECOVERING"
                return FaceUsabilityResult(
                    usable=False,
                    no_face=False,
                    quality_ok=True,
                    occluded=False,
                    occlusion_score=visibility.occlusion_score,
                    occluded_regions=(),
                    visibility_scores=dict(visibility.visibility_scores),
                    region_reasons=dict(visibility.region_reasons),
                    blocking_regions=visibility.blocking_regions,
                    suspicious_regions=visibility.suspicious_regions,
                    quality_status=quality.quality_status,
                    quality_reason=quality.quality_reason,
                    per_region_brightness=dict(quality.per_region_brightness),
                    brightness_uniformity=quality.brightness_uniformity,
                    illumination_score=quality.illumination_score,
                    global_face_brightness=quality.global_face_brightness,
                    shadow_asymmetry=quality.shadow_asymmetry,
                    underexposed_regions=quality.underexposed_regions,
                    overexposed_regions=quality.overexposed_regions,
                    physical_occlusion_score=visibility.occlusion_score,
                    physical_occlusion_regions=visibility.occluded_regions,
                    physical_occlusion_reason=visibility.reason,
                    liveness_skipped_reason="recovering_face_usability",
                    reason="recovering_face_usability",
                    state=self._state,
                    blocked=True,
                    status_override="INSUFFICIENT_EVIDENCE",
                    bbox_detected=True,
                    occlusion_streak=0,
                    quality_streak=0,
                    clear_streak=self._clear_streak,
                )
            self._blocked = False
            self._entered_no_face = False
            self._blocked_mode = None
            self._clear_streak = 0

        self._state = "CLEAR"
        return FaceUsabilityResult(
            usable=True,
            no_face=False,
            quality_ok=True,
            occluded=False,
            occlusion_score=visibility.occlusion_score,
            occluded_regions=(),
            visibility_scores=dict(visibility.visibility_scores),
            region_reasons=dict(visibility.region_reasons),
            blocking_regions=visibility.blocking_regions,
            suspicious_regions=visibility.suspicious_regions,
            quality_status=quality.quality_status,
            quality_reason=quality.quality_reason,
            per_region_brightness=dict(quality.per_region_brightness),
            brightness_uniformity=quality.brightness_uniformity,
            illumination_score=quality.illumination_score,
            global_face_brightness=quality.global_face_brightness,
            shadow_asymmetry=quality.shadow_asymmetry,
            underexposed_regions=quality.underexposed_regions,
            overexposed_regions=quality.overexposed_regions,
            physical_occlusion_score=visibility.occlusion_score,
            physical_occlusion_regions=visibility.occluded_regions,
            physical_occlusion_reason=visibility.reason,
            liveness_skipped_reason="-",
            reason="face_usable",
            state=self._state,
            blocked=False,
            status_override=None,
            bbox_detected=True,
            occlusion_streak=0,
            quality_streak=0,
            clear_streak=self._clear_streak,
        )

    def reset(self) -> None:
        self._quality_streak = 0
        self._occlusion_streak = 0
        self._clear_streak = 0
        self._blocked = False
        self._blocked_mode: Optional[str] = None
        self._entered_no_face = False
        self._state = "CLEAR"

File: web/src/gates/FaceUsabilityGate.ts
// Port of src/gates/face_usability.py — algorithms by @Aysenur15.
//
// Pre-liveness face usability gate. Combines the per-frame quality
// (`IlluminationGate`) and physical-occlusion (`CriticalRegionVisibilityGate`)
// signals into a small state machine that decides whether to feed the
// frame into the heavier liveness pipeline.
//
// State machine states (verbatim from Python):
//   NO_FACE              → bbox missing / frame empty.
//   OCCLUDED_PENDING     → occlusion this frame, streak below confirm threshold.
//   OCCLUDED_CONFIRMED   → occlusion streak ≥ occlusion_confirm_frames.
//   OCCLUDED_NO_FACE     → occlusion streak ≥ no_face_confirm_frames.
//   RECOVERING           → previously blocked, waiting for required_clear_frames.
//   CLEAR                → usable for liveness.
//
// All confirm-frame defaults are preserved verbatim from the source.

import { GateBBox } from "../utils/imageOps";
import { CriticalRegionVisibilityGate, PreviewDetails } from "./CriticalRegionVisibilityGate";
import { IlluminationGate } from "./IlluminationGate";

// Defaults (mirror the module-level constants in face_usability.py).
export const LOW_QUALITY_CONFIRM_FRAMES = 2;
export const OCCLUSION_CONFIRM_FRAMES = 2;
export const NO_FACE_CONFIRM_FRAMES = 6;
export const TEMP_CLEAR_CONFIRM_FRAMES = 1;
export const CLEAR_CONFIRM_FRAMES = 2;

const EYE_STRICT_UNRELIABLE_THRESHOLD = 0.6;

export type UsabilityState =
  | "NO_FACE"
  | "OCCLUDED_PENDING"
  | "OCCLUDED_CONFIRMED"
  | "OCCLUDED_NO_FACE"
  | "RECOVERING"
  | "CLEAR";

export type StatusOverride = "NO_FACE" | "INSUFFICIENT_EVIDENCE" | null;

export interface FaceUsabilityOptions {
  lowQualityConfirmFrames?: number;
  occlusionConfirmFrames?: number;
  noFaceConfirmFrames?: number;
  tempClearConfirmFrames?: number;
  clearConfirmFrames?: number;
}

export interface FaceUsabilityEvaluateOptions {
  /** Optional side-channel preview metrics forwarded to the visibility gate. */
  previewDetails?: PreviewDetails;
  /** Optional blur score for the preview heuristic. */
  blurScore?: number;
}

export interface FaceUsabilityResult {
  usable: boolean;
  noFace: boolean;
  qualityOk: boolean;
  occluded: boolean;
  occlusionScore: number;
  occludedRegions: readonly string[];
  visibilityScores: Record<string, number>;
  regionReasons: Record<string, string>;
  blockingRegions: readonly string[];
  suspiciousRegions: readonly string[];
  qualityStatus: string;
  qualityReason: string;
  perRegionBrightness: Record<string, number>;
  brightnessUniformity: number;
  illuminationScore: number;
  globalFaceBrightness: number;
  shadowAsymmetry: number;
  underexposedRegions: readonly string[];
  overexposedRegions: readonly string[];
  physicalOcclusionScore: number;
  physicalOcclusionRegions: readonly string[];
  physicalOcclusionReason: string;
  livenessSkippedReason: string;
  reason: string;
  state: UsabilityState;
  blocked: boolean;
  statusOverride: StatusOverride;
  bboxDetected: boolean;
  occlusionStreak: number;
  qualityStreak: number;
  clearStreak: number;
}

export class FaceUsabilityGate {
  private readonly qualityGate = new IlluminationGate();
  private readonly visibilityGate = new CriticalRegionVisibilityGate();

  private readonly lowQualityConfirmFrames: number;
  private readonly occlusionConfirmFrames: number;
  private readonly noFaceConfirmFrames: number;
  private readonly tempClearConfirmFrames: number;
  private readonly clearConfirmFrames: number;

  // `qualityStreak` is preserved for parity with the Python state machine,
  // which uses it for the LOW_QUALITY path. The current port treats the
  // illumination signal as informational (rolled into the OCCLUSION state
  // when accompanied by structural failures) and does not advance the
  // streak; the field is left in for upcoming Phase-3 work that wires the
  // LOW_QUALITY confirm path independently.
  private qualityStreak = 0;
  private occlusionStreak = 0;
  private clearStreak = 0;
  private blocked = false;
  private blockedMode: "OCCLUSION" | "LOW_QUALITY" | null = null;
  private enteredNoFace = false;
  private state: UsabilityState = "CLEAR";

  constructor(opts: FaceUsabilityOptions = {}) {
    this.lowQualityConfirmFrames = Math.max(1, Math.floor(opts.lowQualityConfirmFrames ?? LOW_QUALITY_CONFIRM_FRAMES));
    this.occlusionConfirmFrames = Math.max(1, Math.floor(opts.occlusionConfirmFrames ?? OCCLUSION_CONFIRM_FRAMES));
    this.noFaceConfirmFrames = Math.max(
      this.occlusionConfirmFrames,
      Math.floor(opts.noFaceConfirmFrames ?? NO_FACE_CONFIRM_FRAMES),
    );
    this.tempClearConfirmFrames = Math.max(1, Math.floor(opts.tempClearConfirmFrames ?? TEMP_CLEAR_CONFIRM_FRAMES));
    this.clearConfirmFrames = Math.max(1, Math.floor(opts.clearConfirmFrames ?? CLEAR_CONFIRM_FRAMES));
  }

  evaluate(
    frame: ImageData | null,
    faceBbox: GateBBox | null,
    opts: FaceUsabilityEvaluateOptions = {},
  ): FaceUsabilityResult {
    // === No-face short-circuit. ===
    if (!frame || frame.data.length === 0 || !faceBbox) {
      this.qualityStreak = 0;
      this.occlusionStreak = 0;
      this.clearStreak = 0;
      this.blocked = false;
      this.state = "NO_FACE";
      return {
        usable: false,
        noFace: true,
        qualityOk: false,
        occluded: false,
        occlusionScore: 0,
        occludedRegions: [],
        visibilityScores: {},
        regionReasons: {},
        blockingRegions: [],
        suspiciousRegions: [],
        qualityStatus: "-",
        qualityReason: "-",
        perRegionBrightness: {},
        brightnessUniformity: 0,
        illuminationScore: 0,
        globalFaceBrightness: 0,
        shadowAsymmetry: 0,
        underexposedRegions: [],
        overexposedRegions: [],
        physicalOcclusionScore: 0,
        physicalOcclusionRegions: [],
        physicalOcclusionReason: "-",
        livenessSkippedReason: "no_face_detected",
        reason: "no_face_detected",
        state: this.state,
        blocked: true,
        statusOverride: "NO_FACE",
        bboxDetected: false,
        occlusionStreak: 0,
        qualityStreak: 0,
        clearStreak: 0,
      };
    }

    const quality = this.qualityGate.evaluate(frame, faceBbox);
    const visibility = this.visibilityGate.evaluate(frame, faceBbox, {
      previewDetails: opts.previewDetails,
      blurScore: opts.blurScore,
    });

    // Visibility-score thresholds (2026-05-31 relaxation). The original
    // 0.65 nose/mouth bar was calibrated against the Python reference
    // pipeline (480p Python-OpenCV pipeline). On the browser port with
    // MediaPipe FaceLandmarker, a clearly-visible mouth at moderate
    // distance often scores 0.55-0.70 — landmarks are slightly noisy
    // but the region IS visible. At 0.65 the gate was firing
    // "OCCLUDED_NO_FACE / OCCLUDED_CONFIRMED" on 50 %+ of frames in
    // multiple LIVE captures, which pinned the displayed confidence at
    // 34 % even though the LivenessProver was reporting 100/100 PROVEN
    // LIVE — a glaring user-facing inconsistency.
    //
    // Lowered to 0.45 (still well above the ~0.3 "actually occluded"
    // band). Eyes left at 0.6 — eyelid detection is more reliable and
    // a 0.6 score really does indicate eye obstruction.
    const leftEyeScore = visibility.visibilityScores["left_eye"] ?? 1.0;
    const rightEyeScore = visibility.visibilityScores["right_eye"] ?? 1.0;
    const leftEyeVisible = leftEyeScore >= 0.6;
    const rightEyeVisible = rightEyeScore >= 0.6;
    const noseVisible = (visibility.visibilityScores["nose"] ?? 1.0) >= 0.45;
    const mouthVisible = (visibility.visibilityScores["mouth"] ?? 1.0) >= 0.45;
    const lowerFaceVisible = (visibility.visibilityScores["lower_face"] ?? 1.0) >= 0.45;
    const bothEyesUnreliable =
      leftEyeScore < EYE_STRICT_UNRELIABLE_THRESHOLD &&
      rightEyeScore < EYE_STRICT_UNRELIABLE_THRESHOLD;

    const derived: string[] = [...visibility.occludedRegions];
    if (bothEyesUnreliable || (!leftEyeVisible && !rightEyeVisible)) {
      derived.push("left_eye", "right_eye");
    }
    if (!noseVisible) derived.push("nose");
    if (!mouthVisible) derived.push("mouth");
    if (!lowerFaceVisible) derived.push("lower_face");
    const derivedOccludedRegions = Array.from(new Set(derived));

    const lowerFaceRegionsMissing =
      (noseVisible ? 0 : 1) + (mouthVisible ? 0 : 1) + (lowerFaceVisible ? 0 : 1);
    // Structural-occlusion AND-conditions tightened in parallel with the
    // visibility-threshold relaxation above: require a higher occlusionScore
    // to flag, and require more regions missing for the lower-bar trigger.
    // 0.70 occlusionScore is a clearly-bad reading (typical visible-face
    // captures land 0.20-0.45). The original 0.55-0.60 fired too often.
    const structuralOcclusionNow =
      bothEyesUnreliable ||
      (!leftEyeVisible && !rightEyeVisible) ||
      ((!mouthVisible && !lowerFaceVisible) && visibility.occlusionScore >= 0.70) ||
      (lowerFaceRegionsMissing >= 3 && visibility.occlusionScore >= 0.65) ||
      (visibility.occlusionScore >= 0.75 &&
        lowerFaceRegionsMissing >= 2 &&
        (!mouthVisible || !lowerFaceVisible));

    const occludedNow = visibility.isCriticalOccluded || structuralOcclusionNow;

    if (occludedNow) {
      this.qualityStreak = 0;
      this.occlusionStreak += 1;
      this.clearStreak = 0;
      this.blocked = true;
      this.blockedMode = "OCCLUSION";

      let statusOverride: StatusOverride;
      if (this.occlusionStreak >= this.noFaceConfirmFrames) {
        this.enteredNoFace = true;
        this.state = "OCCLUDED_NO_FACE";
        statusOverride = "NO_FACE";
      } else if (this.occlusionStreak >= this.occlusionConfirmFrames) {
        this.state = "OCCLUDED_CONFIRMED";
        statusOverride = "INSUFFICIENT_EVIDENCE";
      } else {
        this.state = "OCCLUDED_PENDING";
        statusOverride = "INSUFFICIENT_EVIDENCE";
      }
      const suspicious = uniqueExcept(
        [...visibility.suspiciousRegions, ...derivedOccludedRegions],
        derivedOccludedRegions,
      );
      return {
        usable: false,
        noFace: true,
        qualityOk: true,
        occluded: true,
        occlusionScore: visibility.occlusionScore,
        occludedRegions: derivedOccludedRegions,
        visibilityScores: { ...visibility.visibilityScores },
        regionReasons: { ...visibility.regionReasons },
        blockingRegions: derivedOccludedRegions,
        suspiciousRegions: suspicious,
        qualityStatus: quality.qualityStatus,
        qualityReason: quality.qualityReason,
        perRegionBrightness: { ...quality.perRegionBrightness },
        brightnessUniformity: quality.brightnessUniformity,
        illuminationScore: quality.illuminationScore,
        globalFaceBrightness: quality.globalFaceBrightness,
        shadowAsymmetry: quality.shadowAsymmetry,
        underexposedRegions: [...quality.underexposedRegions],
        overexposedRegions: [...quality.overexposedRegions],
        physicalOcclusionScore: visibility.occlusionScore,
        physicalOcclusionRegions: derivedOccludedRegions,
        physicalOcclusionReason: visibility.isCriticalOccluded
          ? visibility.reason
          : "structural_face_region_occluded",
        livenessSkippedReason: "critical_face_region_occluded",
        reason: "critical_face_region_occluded",
        state: this.state,
        blocked: true,
        statusOverride,
        bboxDetected: true,
        occlusionStreak: this.occlusionStreak,
        qualityStreak: 0,
        clearStreak: 0,
      };
    }

    // === Clear branch. ===
    this.occlusionStreak = 0;
    this.qualityStreak = 0;
    if (this.blocked) {
      this.clearStreak += 1;
      const requiredClearFrames =
        this.enteredNoFace || this.blockedMode === "LOW_QUALITY"
          ? this.clearConfirmFrames
          : this.tempClearConfirmFrames;
      if (this.clearStreak < requiredClearFrames) {
        this.state = "RECOVERING";
        return {
          usable: false,
          noFace: false,
          qualityOk: true,
          occluded: false,
          occlusionScore: visibility.occlusionScore,
          occludedRegions: [],
          visibilityScores: { ...visibility.visibilityScores },
          regionReasons: { ...visibility.regionReasons },
          blockingRegions: visibility.blockingRegions,
          suspiciousRegions: visibility.suspiciousRegions,
          qualityStatus: quality.qualityStatus,
          qualityReason: quality.qualityReason,
          perRegionBrightness: { ...quality.perRegionBrightness },
          brightnessUniformity: quality.brightnessUniformity,
          illuminationScore: quality.illuminationScore,
          globalFaceBrightness: quality.globalFaceBrightness,
          shadowAsymmetry: quality.shadowAsymmetry,
          underexposedRegions: [...quality.underexposedRegions],
          overexposedRegions: [...quality.overexposedRegions],
          physicalOcclusionScore: visibility.occlusionScore,
          physicalOcclusionRegions: visibility.occludedRegions,
          physicalOcclusionReason: visibility.reason,
          livenessSkippedReason: "recovering_face_usability",
          reason: "recovering_face_usability",
          state: this.state,
          blocked: true,
          statusOverride: "INSUFFICIENT_EVIDENCE",
          bboxDetected: true,
          occlusionStreak: 0,
          qualityStreak: 0,
          clearStreak: this.clearStreak,
        };
      }
      this.blocked = false;
      this.enteredNoFace = false;
      this.blockedMode = null;
      this.clearStreak = 0;
    }

    // CLEAR — usable.
    this.state = "CLEAR";
    // Touch the unused low_quality_confirm field + quality_streak for
    // noUnusedLocals; these exist for parity with the Python signature and
    // are reserved for the LOW_QUALITY confirm path that Phase 3 wires.
    void this.lowQualityConfirmFrames;
    void this.qualityStreak;
    return {
      usable: true,
      noFace: false,
      qualityOk: true,
      occluded: false,
      occlusionScore: visibility.occlusionScore,
      occludedRegions: [],
      visibilityScores: { ...visibility.visibilityScores },
      regionReasons: { ...visibility.regionReasons },
      blockingRegions: visibility.blockingRegions,
      suspiciousRegions: visibility.suspiciousRegions,
      qualityStatus: quality.qualityStatus,
      qualityReason: quality.qualityReason,
      perRegionBrightness: { ...quality.perRegionBrightness },
      brightnessUniformity: quality.brightnessUniformity,
      illuminationScore: quality.illuminationScore,
      globalFaceBrightness: quality.globalFaceBrightness,
      shadowAsymmetry: quality.shadowAsymmetry,
      underexposedRegions: [...quality.underexposedRegions],
      overexposedRegions: [...quality.overexposedRegions],
      physicalOcclusionScore: visibility.occlusionScore,
      physicalOcclusionRegions: visibility.occludedRegions,
      physicalOcclusionReason: visibility.reason,
      livenessSkippedReason: "-",
      reason: "face_usable",
      state: this.state,
      blocked: false,
      statusOverride: null,
      bboxDetected: true,
      occlusionStreak: 0,
      qualityStreak: 0,
      clearStreak: this.clearStreak,
    };
  }

  /** Reset all streak counters. */
  reset(): void {
    this.qualityStreak = 0;
    this.occlusionStreak = 0;
    this.clearStreak = 0;
    this.blocked = false;
    this.blockedMode = null;
    this.enteredNoFace = false;
    this.state = "CLEAR";
    this.visibilityGate.reset();
  }
}

function uniqueExcept(items: readonly string[], exclude: readonly string[]): string[] {
  const excl = new Set(exclude);
  const seen = new Set<string>();
  const out: string[] = [];
  for (const v of items) {
    if (excl.has(v)) continue;
    if (seen.has(v)) continue;
    seen.add(v);
    out.push(v);
  }
  return out;
}
